File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

"""
input: linked list head 
output: linked list head (with removed element)

Plan: 
High Level: Create a pointer at the head and go through the list all the way to the end to get the total number of nodes.
Then with the total number of nodes, calculate how many steps it would take to get the the nth from last node using the total - n. 
total - n = index we are looking to remove 
ex 1: 4 - 2 = 2 (remove element at index 2)
ex 3: 2 - 2 = 0 (remove lement at index 0)
"""

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # find the length of the list
        curr = head
        length = 0

        while curr: 
            length += 1
            curr = curr.next
        
        print(f"total length = {length}")

        indexToRemove = length - n

        if n > length: 
            return head

        temp = ListNode(0, head)
        prev = temp
        curr = head 
        currIndex = 0

        while curr:
            if currIndex == indexToRemove:
                prev.next = curr.next
                break
            
            prev = curr
            curr = curr.next
            currIndex += 1

        return temp.next

            

File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode curr = head;
        ListNode prev = null;
        ListNode tmp = null;

        // while current pointer is not null, set temp as current and move current
        // temp's next is equal to the previous
        while(curr != null) {
            tmp = curr;
            curr = curr.next;
            tmp.next = prev;
            prev = tmp;
        }

        // once curr is at null, prev points to the last node of the list
        head = prev;

        return head;
    }
}
